File: binary search/74. Search a 2D Matrix/74. Search a 2D Matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  const m = matrix.length;
  const n = matrix[0].length;
  let start = 0;
  let end = m * n - 1;
  let mid;
  while (start <= end) {
    mid = start + ((end - start) >> 1);
    const num = matrix[Math.floor(mid / n)][mid % n];
    if (num === target) return true;
    else num > target ? (end = mid - 1) : (start = mid + 1);
  }
  return false;
};

File: binary search/74. Search a 2D Matrix.java
class Solution {
  public boolean searchMatrix(int[][] matrix, int target) {
    int rows = matrix.length;
    int cols = matrix[0].length;
    int l = 0;
    int r = rows * cols - 1;
    while (l <= r) {
      int mid = l + (r - l) / 2;
      if (matrix[mid / cols][mid % cols] == target)
        return true;
      else if (matrix[mid / cols][mid % cols] < target)
        l = mid + 1;
      else
        r = mid - 1;
    }
    return false;
  }
}